File: Code/frontend/src/types/topic.ts
export type TopicType = "LECTURE" | "LAB" | "SEMINAR" | "PROJECT" | "SELF_STUDY"

export interface Topic {
  id: number
  syllabusId: number
  weekNumber: number
  orderInWeek?: number
  name: string
  nameVn?: string
  teachingHours?: number
  labHours?: number
  selfStudyHours?: number
  topicType?: TopicType | string
  teachingMethod?: string
  learningActivity?: string
  notes?: string
}

export interface CreateTopicRequest {
  syllabusId: number
  weekNumber: number
  orderInWeek?: number
  name: string
  nameVn?: string
  teachingHours?: number
  labHours?: number
  selfStudyHours?: number
  topicType?: string
  teachingMethod?: string
  learningActivity?: string
  notes?: string
}

export const TOPIC_TYPES: { value: TopicType; label: string }[] = [
  { value: "LECTURE", label: "Lý thuyết (Lecture)" },
  { value: "LAB", label: "Thực hành (Lab)" },
  { value: "SEMINAR", label: "Seminar" },
  { value: "PROJECT", label: "Dự án (Project)" },
  { value: "SELF_STUDY", label: "Tự học (Self Study)" },
]

File: Code/backend/curriculum/src/main/java/com/scse/curriculum/topic/entity/Topic.java
package com.scse.curriculum.topic.entity;

import com.scse.curriculum.syllabus.entity.Syllabus;
import jakarta.persistence.*;
import lombok.*;

@Entity
@Table(name = "topic")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Topic {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "syllabus_id")
    private Syllabus syllabus;

    @Column(name = "week_number")
    private Integer weekNumber;

    @Column(name = "order_in_week")
    private Integer orderInWeek;

    @Column(nullable = false)
    private String name;

    @Column(name = "name_vn")
    private String nameVn;

    @Column(name = "teaching_hours")
    @Builder.Default
    private Integer teachingHours = 3;

    @Column(name = "lab_hours")
    @Builder.Default
    private Integer labHours = 0;

    @Column(name = "self_study_hours")
    @Builder.Default
    private Integer selfStudyHours = 0;

    @Enumerated(EnumType.STRING)
    @Column(name = "topic_type")
    @Builder.Default
    private TopicType topicType = TopicType.LECTURE;

    @Column(name = "teaching_method",
            columnDefinition = "TEXT")
    private String teachingMethod;

    @Column(name = "learning_activity",
            columnDefinition = "TEXT")
    private String learningActivity;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @Column(name = "assessments",
        columnDefinition = "TEXT")
private String assessments;


@Column(name = "resources",
        columnDefinition = "TEXT")
private String resources;

@PrePersist
private void initializeRequiredValues() {
    if (weekNumber == null) weekNumber = 1;
    if (orderInWeek == null) orderInWeek = 1;
    if (teachingHours == null) teachingHours = 3;
    if (labHours == null) labHours = 0;
    if (selfStudyHours == null) selfStudyHours = 0;
    if (topicType == null) topicType = TopicType.LECTURE;
}
}
